dbt/include/dremio/macros/materializations/incremental/incremental.sql

```diff
@@ -14,64 +14,96 @@ limitations under the License.*/
-                                                database=database,
-                                                type='table') -%}
-
-  {%- set tmp_identifier = model['name'] + '__dbt_tmp' -%}
-  {%- set tmp_relation = adapter.get_relation(database=database, schema=schema, identifier=tmp_identifier) -%}
-  {%- set target_tmp_relation = api.Relation.create(identifier=tmp_identifier,
-                                                schema=schema,
-                                                database=database,
-                                                type='table') -%}
-
-  {%- set raw_strategy = config.get('incremental_strategy', validator=validation.any[basestring]) or 'append' -%}
-  {%- set raw_file_format = config.get('format', validator=validation.any[basestring]) or 'iceberg' -%}
-  {%- set file_format = dbt_dremio_validate_get_file_format(raw_file_format) -%}
-  {%- set strategy = dbt_dremio_validate_get_incremental_strategy(raw_strategy, file_format) -%}
+  -- relations
+  {%- set existing_relation = load_cached_relation(this) -%} 
+  {%- set target_relation = this.incorporate(type='table') -%}
+  {%- set temp_relation = make_temp_relation(target_relation) -%}
+  {%- set intermediate_relation = make_intermediate_relation(target_relation) -%}
+  {%- set backup_relation_type = 'table' if existing_relation is none else existing_relation.type -%}
+  {%- set backup_relation = make_backup_relation(target_relation, backup_relation_type) -%}
+
+  -- configs
   {%- set unique_key = config.get('unique_key', validator=validation.any[list, basestring]) -%}
   {%- set full_refresh_mode = (should_full_refresh()) -%}
-  {%- set raw_on_schema_change = config.get('on_schema_change', validator=validation.any[basestring]) or 'ignore' -%}
   {%- set on_schema_change = incremental_validate_on_schema_change(raw_on_schema_change) -%}
-  {%- set full_refresh_mode = (should_full_refresh()) -%}
+
+  -- the temp_ and backup_ relations should not already exist in the database; get_relation
+  -- will return None in that case. Otherwise, we get a relation that we can drop
+  -- later, before we try to use this name for the current operation. This has to happen before
+  -- BEGIN, in a separate transaction
+  {%- set preexisting_intermediate_relation = load_cached_relation(intermediate_relation)-%}
+  {%- set preexisting_backup_relation = load_cached_relation(backup_relation) -%}
+  -- grab current tables grants config for comparison later on
   {% set grant_config = config.get('grants') %}
+  {{ drop_relation_if_exists(preexisting_intermediate_relation) }}
+  {{ drop_relation_if_exists(preexisting_backup_relation) }}
+
+  {{ run_hooks(pre_hooks, inside_transaction=False) }}
+ 
+  -- `BEGIN` happens here:
+  {{ run_hooks(pre_hooks, inside_transaction=True) }}
 
-  {{ run_hooks(pre_hooks) }}
+  {% set to_drop = [] %}
 
-  {% if old_relation is none %}
-    {% set build_sql = create_table_as(False, target_relation, external_query(sql)) %}

```

Comment:
The old build_sql  had the sql wrapped in an external_query macro. Is there a reason we removed this?

Reply:
We really should be using external_query. I will change this. Thanks!